### PR TITLE
Codegen for bitwise and, or, xor, and not 

### PR DESCRIPTION
**Pull request by steventk-g**
## PyTorch PR

https://github.com/pytorch/pytorch/pull/82617

## LazyIr.h
```
class BitwiseAndTensor : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::bitwise_and);
  }

  BitwiseAndTensor(const torch::lazy::Value& self, const torch::lazy::Value& other, std::vector<torch::lazy::Shape>&& shapes)
      : XlaNode(torch::lazy::OpKind(at::aten::bitwise_and),
              {self, other}, std::move(shapes),
              [&]() { return BitwiseAndTensorOutputShape(self, other); },
              /* num_outputs */ 1,
              torch::lazy::MHash())
  {
    
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    
    return ss.str();
  }

  

  bool CanBeReused(const torch::lazy::Value& self, const torch::lazy::Value& other) const {
    return false;
    }

  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;

  
  

};

class BitwiseNot : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::bitwise_not);
  }

  BitwiseNot(const torch::lazy::Value& self, std::vector<torch::lazy::Shape>&& shapes)
      : XlaNode(torch::lazy::OpKind(at::aten::bitwise_not),
              {self}, std::move(shapes),
              [&]() { return BitwiseNotOutputShape(self); },
              /* num_outputs */ 1,
              torch::lazy::MHash())
  {
    
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    
    return ss.str();
  }

  

  bool CanBeReused(const torch::lazy::Value& self) const {
    return false;
    }

  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;

  
  

};

class BitwiseOrTensor : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::bitwise_or);
  }

  BitwiseOrTensor(const torch::lazy::Value& self, const torch::lazy::Value& other, std::vector<torch::lazy::Shape>&& shapes)
      : XlaNode(torch::lazy::OpKind(at::aten::bitwise_or),
              {self, other}, std::move(shapes),
              [&]() { return BitwiseOrTensorOutputShape(self, other); },
              /* num_outputs */ 1,
              torch::lazy::MHash())
  {
    
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    
    return ss.str();
  }

  

  bool CanBeReused(const torch::lazy::Value& self, const torch::lazy::Value& other) const {
    return false;
    }

  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;

  
  

};

class BitwiseXorTensor : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::bitwise_xor);
  }

  BitwiseXorTensor(const torch::lazy::Value& self, const torch::lazy::Value& other, std::vector<torch::lazy::Shape>&& shapes)
      : XlaNode(torch::lazy::OpKind(at::aten::bitwise_xor),
              {self, other}, std::move(shapes),
              [&]() { return BitwiseXorTensorOutputShape(self, other); },
              /* num_outputs */ 1,
              torch::lazy::MHash())
  {
    
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    
    return ss.str();
  }

  

  bool CanBeReused(const torch::lazy::Value& self, const torch::lazy::Value& other) const {
    return false;
    }

  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;

  
  

};
```

## XLANativeFunctions.cpp

```
    at::Tensor XLANativeFunctions::bitwise_and(const at::Tensor & self, const at::Tensor & other) {
        
        XLA_FN_COUNTER("xla::");
        auto common_device = torch_xla::bridge::GetXlaDevice(self, other);
        TORCH_INTERNAL_ASSERT(common_device);
        
        torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
        torch_xla::XLATensorPtr lazy_other = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(other, *common_device);
        torch::lazy::NodePtr node = torch::lazy::ReuseNode<BitwiseAndTensor>(lazy_self->GetIrValue(), lazy_other->GetIrValue());
        if (!node) {
                    auto self_meta = to_meta(self);
        auto other_meta = to_meta(other);
        auto out_meta = at::meta::bitwise_and(self_meta, other_meta);
        
std::vector<torch::lazy::Shape> shapes{torch::lazy::Shape(out_meta.scalar_type(), out_meta.sizes().vec())};
            TORCH_INTERNAL_ASSERT(shapes.size() == 1);
            if(torch::lazy::symbolicShapeEnabled()){
                std::vector<torch::jit::IValue> inputs = { self, other };
                const char* schema_str = "aten::bitwise_and.Tensor(Tensor self, Tensor other) -> Tensor";
                applySymbolicShapesOnLT(schema_str, inputs, shapes);
            }
        
            node = torch::lazy::MakeNode<BitwiseAndTensor>(lazy_self->GetIrValue(), lazy_other->GetIrValue(), std::move(shapes));
            CacheNode(node);
        }
        
        auto result = torch_xla::bridge::AtenFromXlaTensor(
                torch_xla::XLATensor::Create(std::move(node), *common_device));
        return result;
    };

    
    at::Tensor XLANativeFunctions::bitwise_not(const at::Tensor & self) {
        
        XLA_FN_COUNTER("xla::");
        auto common_device = torch_xla::bridge::GetXlaDevice(self);
        TORCH_INTERNAL_ASSERT(common_device);
        
        torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
        torch::lazy::NodePtr node = torch::lazy::ReuseNode<BitwiseNot>(lazy_self->GetIrValue());
        if (!node) {
                    auto self_meta = to_meta(self);
        auto out_meta = at::meta::bitwise_not(self_meta);
        
std::vector<torch::lazy::Shape> shapes{torch::lazy::Shape(out_meta.scalar_type(), out_meta.sizes().vec())};
            TORCH_INTERNAL_ASSERT(shapes.size() == 1);
            if(torch::lazy::symbolicShapeEnabled()){
                std::vector<torch::jit::IValue> inputs = { self };
                const char* schema_str = "aten::bitwise_not(Tensor self) -> Tensor";
                applySymbolicShapesOnLT(schema_str, inputs, shapes);
            }
        
            node = torch::lazy::MakeNode<BitwiseNot>(lazy_self->GetIrValue(), std::move(shapes));
            CacheNode(node);
        }
        
        auto result = torch_xla::bridge::AtenFromXlaTensor(
                torch_xla::XLATensor::Create(std::move(node), *common_device));
        return result;
    };

    
    at::Tensor XLANativeFunctions::bitwise_or(const at::Tensor & self, const at::Tensor & other) {
        
        XLA_FN_COUNTER("xla::");
        auto common_device = torch_xla::bridge::GetXlaDevice(self, other);
        TORCH_INTERNAL_ASSERT(common_device);
        
        torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
        torch_xla::XLATensorPtr lazy_other = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(other, *common_device);
        torch::lazy::NodePtr node = torch::lazy::ReuseNode<BitwiseOrTensor>(lazy_self->GetIrValue(), lazy_other->GetIrValue());
        if (!node) {
                    auto self_meta = to_meta(self);
        auto other_meta = to_meta(other);
        auto out_meta = at::meta::bitwise_or(self_meta, other_meta);
        
std::vector<torch::lazy::Shape> shapes{torch::lazy::Shape(out_meta.scalar_type(), out_meta.sizes().vec())};
            TORCH_INTERNAL_ASSERT(shapes.size() == 1);
            if(torch::lazy::symbolicShapeEnabled()){
                std::vector<torch::jit::IValue> inputs = { self, other };
                const char* schema_str = "aten::bitwise_or.Tensor(Tensor self, Tensor other) -> Tensor";
                applySymbolicShapesOnLT(schema_str, inputs, shapes);
            }
        
            node = torch::lazy::MakeNode<BitwiseOrTensor>(lazy_self->GetIrValue(), lazy_other->GetIrValue(), std::move(shapes));
            CacheNode(node);
        }
        
        auto result = torch_xla::bridge::AtenFromXlaTensor(
                torch_xla::XLATensor::Create(std::move(node), *common_device));
        return result;
    };

    
    at::Tensor XLANativeFunctions::bitwise_xor(const at::Tensor & self, const at::Tensor & other) {
        
        XLA_FN_COUNTER("xla::");
        auto common_device = torch_xla::bridge::GetXlaDevice(self, other);
        TORCH_INTERNAL_ASSERT(common_device);
        
        torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
        torch_xla::XLATensorPtr lazy_other = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(other, *common_device);
        torch::lazy::NodePtr node = torch::lazy::ReuseNode<BitwiseXorTensor>(lazy_self->GetIrValue(), lazy_other->GetIrValue());
        if (!node) {
                    auto self_meta = to_meta(self);
        auto other_meta = to_meta(other);
        auto out_meta = at::meta::bitwise_xor(self_meta, other_meta);
        
std::vector<torch::lazy::Shape> shapes{torch::lazy::Shape(out_meta.scalar_type(), out_meta.sizes().vec())};
            TORCH_INTERNAL_ASSERT(shapes.size() == 1);
            if(torch::lazy::symbolicShapeEnabled()){
                std::vector<torch::jit::IValue> inputs = { self, other };
                const char* schema_str = "aten::bitwise_xor.Tensor(Tensor self, Tensor other) -> Tensor";
                applySymbolicShapesOnLT(schema_str, inputs, shapes);
            }
        
            node = torch::lazy::MakeNode<BitwiseXorTensor>(lazy_self->GetIrValue(), lazy_other->GetIrValue(), std::move(shapes));
            CacheNode(node);
        }
        
        auto result = torch_xla::bridge::AtenFromXlaTensor(
                torch_xla::XLATensor::Create(std::move(node), *common_device));
        return result;
    };
```

## Testing
Built with `BUILD_CPP_TESTS=0 python setup.py install` in docker container on cloud top.

In python shell, ran:
```
>>> torch.bitwise_and(torch.tensor([-1, -2, 3], dtype=torch.int8, device=xm.xla_device()), torch.tensor([1, 0, 3], dtype=torch.int8, device=xm.xla_device()))
tensor([1, 0, 3], device='xla:0', dtype=torch.int8)
>>> torch.bitwise_or(torch.tensor([-1, -2, 3], dtype=torch.int8, device=xm.xla_device()), torch.tensor([1, 0, 3], dtype=torch.int8, device=xm.xla_device()))
tensor([-1, -2,  3], device='xla:0', dtype=torch.int8)
>>> torch.bitwise_xor(torch.tensor([-1, -2, 3], dtype=torch.int8, device=xm.xla_device()), torch.tensor([1, 0, 3], dtype=torch.int8, device=xm.xla_device()))
tensor([-2, -2,  0], device='xla:0', dtype=torch.int8)
>>> torch.bitwise_not(torch.tensor([-1, -2, 3], dtype=torch.int8, device=xm.xla_device()))
tensor([ 0,  1, -4], device='xla:0', dtype=torch.int8)
>>> torch.bitwise_and(torch.tensor([True, True, False], device=xm.xla_device()), torch.tensor([False, True, False], device=xm.xla_device()))
tensor([False,  True, False], device='xla:0')
>>> torch.bitwise_or(torch.tensor([True, True, False], device=xm.xla_device()), torch.tensor([False, True, False], device=xm.xla_device()))
tensor([ True,  True, False], device='xla:0')
>>> torch.bitwise_xor(torch.tensor([True, True, False], device=xm.xla_device()), torch.tensor([False, True, False], device=xm.xla_device()))
tensor([ True, False, False], device='xla:0')
>>> torch.bitwise_not(torch.tensor([True, True, False], device=xm.xla_device()))
tensor([False, False,  True], device='xla:0')
```

### CODE REVIEW

File: xla_native_functions.yaml [hunk 1]
```diff
@@ -118,13 +122,6 @@ supported:
   - bernoulli_.float
   - bernoulli_.Tensor
   - binary_cross_entropy_with_logits
-  - bitwise_and.Scalar
```
Comment (JackCaoG) [inline xla_native_functions.yaml:121]: any reason we remove all scalar version of overwrites? Did you verify that when user call `torch.bitwise_and` for `tensor, scalar` it got dispatched to tensor version?

Comment (steventk-g) [reply xla_native_functions.yaml:121]: Yes, testing with the following shows that `xla::bitwise_and` gets executed on scalar operations.
```
>>> import torch; import torch_xla
>>> torch.bitwise_and(torch.tensor(100,device='xla:0'), 10)
tensor(0, device='xla:0')
>>> import torch_xla.debug.metrics as met; print(met.metrics_report())
```

File: torch_xla/csrc/tensor_methods.cpp [hunk 2]
```diff
@@ -864,52 +863,39 @@ void XLATensor::bernoulli_(XLATensorPtr& input,
...
 
-XLATensorPtr XLATensor::bitwise_and(const XLATensorPtr& input,
-                                    const at::Scalar& other) {
-  CheckIsIntegralOrPred(input->shape(), "__and__");
-  torch::lazy::Value constant =
-      GetIrValueForScalar(other, input->shape(), input->GetDevice());
-  return input->CreateFrom(BitwiseAnd(input->GetIrValue(), constant));
-}
-
 XLATensorPtr XLATensor::bitwise_and(const XLATensorPtr& input,
```
Comment (JackCaoG) [inline torch_xla/csrc/tensor_methods.cpp:866]: we should remove `XLATensor::bitwise_xxx` since they are not needed anymore

Comment (steventk-g) [reply torch_xla/csrc/tensor_methods.cpp]: Done

File: torch_xla/csrc/ops/ops_xla_shape_fn.cpp [hunk 3]
```diff
@@ -193,6 +193,25 @@ xla::Shape BinaryCrossEntropyBackwardOutputShape(
   return InferOutputShape(shapes, lower_for_shape_fn);
 }
 
+xla::Shape BitwiseAndTensorOutputShape(const torch::lazy::Value& input,
+                                       const torch::lazy::Value& other) {
+  return GetXlaShape(input);
```
Comment (JackCaoG) [inline torch_xla/csrc/ops/ops_xla_shape_fn.cpp:198]: This doesn't seem right, for example
```
torch.bitwise_and(torch.tensor([-1], dtype=torch.int8), torch.tensor([1, 0, 3], dtype=torch.int8))

tensor([1, 0, 3], dtype=torch.int8)
```
You want to reuse lowering logic above(maybe make them helper functions).

Comment (steventk-g) [reply torch_xla/csrc/ops/ops_xla_shape_fn.cpp]: Done

**Review by JackCaoG — APPROVED** [review]
Thanks!

**Comment by JackCaoG** [comment]
@steventk-g Feel free to merge when all test passed

**Review by wonjoolee95 — APPROVED** [review]
👍

**Pull request merged**